magma/smart/smart_bits.py

```diff
@@ -0,0 +1,526 @@
+import abc
+import copy
+import enum
+import inspect
+import operator
+
+from magma.bits import Bits, BitsMeta, SInt
+from magma.conversions import uint, bits, sint
+from magma.conversions import concat as bits_concat
+from magma.debug import debug_wire
+from magma.primitives.reduce import reduce
+from magma.protocol_type import MagmaProtocolMeta, MagmaProtocol
+
+
+def _is_int(value):
+    try:
+        int(value)
+    except (ValueError, TypeError):
+        return False
+    return True
+
+
+class Context:
+    def __init__(self, assignee, root):
+        self.assignee = assignee
+        self.root = root
+
+    def max_width(self):
+
+        def _visit(node):
+            if hasattr(node, "_width_"):
+                return node._width_
+            if isinstance(node, _SmartBitsExpr):
+                return len(node.bits)
+            return max(_visit(arg) for arg in node.args)
+
+        max_width = _visit(self.root)
+        if self.assignee is not None:
+            max_width = max(max_width, len(self.assignee.bits))
+        return max_width
+
+    def __str__(self):
+        return f"Context(lhs={repr(self.assignee)}, rhs={repr(self.root)})"
+
+
+class _SmartExprMeta(MagmaProtocolMeta):
+    pass
+
+
+class _SmartExpr(MagmaProtocol, metaclass=_SmartExprMeta):
+    @abc.abstractmethod
+    def resolve(self, context):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def eval(self):
+        raise NotImplementedError()
+
+    # Binary arithmetic operators.
+    def __add__(self, other: '_SmartExpr'):
+        if not isinstance(other, _SmartExpr):
```

Comment:
Using `not isinstance(other, type(self))` make this more robust with inheritance

Reply:
the problem with that is we can call the operator on two separate sub-classes, so

```python
class Sub1(Super): ...
class Sub2(Super): ...

Sub1() + Sub2()  # not isinstance(Sub2(), type(Sub1())), but we want this to work
```

Reply:
Do we want that to work?  What is the return type?

Reply:
Further what is the type of `Sub1() + Sub1()`

Reply:
If the issue is the mixing of unsigned and signed I would advocate checking for that case explicitly like have a way to get the non-signed (i.e. the type sans signedness information) from a SmartBits type and checking if other is an instance of that.

Reply:
The issue is not of signed-ness; rather its just a matter of the implementation design. That is `_SmartExpr` is an abstract base class...each "ast node" inherits from that (just colloquially calling them AST nodes, in reality they're just these staged values, but each type of op has it's own node, e.g. AddNode, ShiftNode, etc.). So we define these operators on the base class and get them for free. It's kind of like:

```python
class Node:
    def __add__(self, other):
        assert isinstance(other, Node)
        return AddNode(self, other)

   def __sub__(self, other):
        assert isinstance(other, Node)
        return SubNode(self, other)

class AddNode(Node): ...
class SubNode(Node): ...

class Leaf(Node): pass

x = Leaf()
y = Leaf()
z = (x + y) - (x + x - y) + ...  # whatever you want to do on x and y
```

All the intermediates, such as `(x + y)`, are indirectly instances of `Node` base cls, but we they could be any specialized node type.

Reply:
Resolving for now; if you think there is better design, would be good to discuss. I'm not tied to this way, but it seems to be working.

Reply:
Gotcha, it might be possible to have a subtype of SmartBits automatically generate its own ast type so that it could be safely subtyped.  Let me think about it.